src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsSolver.cpp

```diff
@@ -128,19 +135,36 @@ void MultiphasePoromechanicsSolver::assembleSystem( real64 const GEOSX_UNUSED_PA
-                                                                                    flowSolver()->numFluidComponents(),
-                                                                                    flowSolver()->numFluidPhases(),
-                                                                                    FlowSolverBase::viewKeyStruct::fluidNamesString() );
-
+    if( m_isThermal )
+    {
+      poromechanicsMaxForce =
+        assemblyLaunch< constitutive::PorousSolidBase,
+                        thermalPoromechanicsKernels::ThermalMultiphasePoromechanicsKernelFactory >( mesh,
+                                                                                                    dofManager,
+                                                                                                    regionNames,
+                                                                                                    viewKeyStruct::porousMaterialNamesString(),
+                                                                                                    localMatrix,
+                                                                                                    localRhs,
+                                                                                                    flowDofKey,
+                                                                                                    flowSolver()->numFluidComponents(),
+                                                                                                    flowSolver()->numFluidPhases(),
+                                                                                                    FlowSolverBase::viewKeyStruct::fluidNamesString() );
+    }
+    else
+    {
+      poromechanicsMaxForce =
+        assemblyLaunch< constitutive::PorousSolidBase,
+                        poromechanicsKernels::MultiphasePoromechanicsKernelFactory >( mesh,
+                                                                                      dofManager,
+                                                                                      regionNames,
+                                                                                      viewKeyStruct::porousMaterialNamesString(),
+                                                                                      localMatrix,
+                                                                                      localRhs,
+                                                                                      flowDofKey,
+                                                                                      flowSolver()->numFluidComponents(),
+                                                                                      flowSolver()->numFluidPhases(),
+                                                                                      FlowSolverBase::viewKeyStruct::fluidNamesString() );
+    }
```

Comment:
Does this compile with our set of compilers?
```suggestion
auto args = std::make_tuple{
  mesh,
  dofManager,
  regionNames,
  viewKeyStruct::porousMaterialNamesString(),
  localMatrix,
  localRhs,
  flowDofKey,
  flowSolver()->numFluidComponents(),
  flowSolver()->numFluidPhases(),
  FlowSolverBase::viewKeyStruct::fluidNamesString()
};
poromechanicsMaxForce = m_isThermal ?
  std::apply(
    assemblyLaunch<
      constitutive::PorousSolidBase,
      thermalPoromechanicsKernels::ThermalMultiphasePoromechanicsKernelFactory
  >, args ):
  std::apply(
    assemblyLaunch<
      constitutive::PorousSolidBase,
      poromechanicsKernels::MultiphasePoromechanicsKernelFactory
  >, args )
```
I do not know if it's worth it. It's just a handmade dispatch, we surely have a bunch in `geosx`.

Reply:
Nice! I have been looking for something like that for a while, I did not know it existed. Let me give it a try

Reply:
Mind the copy/reference issues.
You may want to play with `std::forward_as_tuple`, `std::tie`, `std::make_tuple`, and `std::ref` or `std::cref`. Looks like `std::forward_as_tuple` may be the most appropriate choice, but this needs to be carefully checked.

Reply:
@TotoGaz before I spend time fixing my compilation with `std::apply`, just a quick question: is `std::apply` supported by the version of C++ that we use?

I am asking because I get:
```
/g/g90/hamon1/geosx/GEOSX/src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsSolver.cpp:187:12: error:                                                                
      no member named 'apply' in namespace 'std'                                                                                                                                         
      std::apply( assemblyLaunch<
```
but I may have other issues as well.

Reply:
Hmm [not sure](https://en.cppreference.com/w/cpp/utility/apply) 🥺 
It looks like it's 17. We'll be using 17 soon, but not yet 🤷

Reply:
In another PR https://github.com/GEOSX/GEOSX/pull/2312#discussion_r1125553951 we also have use case for a `std::apply`-like function, so we will probably add a hand-rolled implementation (it's quite simple if we ignore `noexcept` and complexities of `std::invoke`). We could just converge on one implementation and put it in `Utilities` somewhere, to accompany `forEachArgInTuple`.

Reply:
Ok great, thanks very much @XL64